### ISSUE TITLE
Wrong initial RAM contents in some machines

### ISSUE THREAD

**Issue by theNestruo** [openMSX/openMSX#1249]
Hi!

It seems that several machines have a wrong or missing `<initialContent />` element in their definitions. Or, at least, there are machines out there that do not match the initial RAM content defined in OpenMSX.

For example, my Philips NMS 8250 does not match the `<!-- (chr(0) + chr(255)) * 128 + (chr(255)+ chr(0)) * 128) -->` pattern described in [its definition file](https://github.com/openMSX/openMSX/blob/master/share/machines/Philips_NMS_8250.xml):
![NMS8250](https://user-images.githubusercontent.com/12213477/80823111-47175980-8bdc-11ea-815a-19f8cf706769.jpg)

The VG-8235 of a friend:
![VG8235](https://user-images.githubusercontent.com/12213477/80823819-914d0a80-8bdd-11ea-9c41-3c2c14e2bff7.jpg)

However, the pattern currently defined for those two machines doesn't seem arbitrary at all... so I don't know if it depends on the RAM chips maker, model, serial number, or any other variable...

---

I guess a small BASIC line can be used to inspect the RAM: `WIDTH32:FORA=&HE000TO&HE0FF:?RIGHT$("0"+HEX$(PEEK(A)),2);:NEXT` (at least in a tentatively manner), so in order to gather information from The Internet™ I [tweeted](https://twitter.com/NestorSancho/status/1256237405880885249) it.

I had some replies. First, the discrepancied:

* A [Panasonic WSX](https://twitter.com/ibannieto/status/1256257128471375872) shows a `00FF00FF...` pattern, not present in [its definition file](https://github.com/openMSX/openMSX/blob/master/share/machines/Panasonic_FS-A1WSX.xml) _Note: I should have included a few more bytes, because now I'm not sure if the pattern is as is, or if it is negated for the next 256 bytes, as in the NMS8245_
* Another [Philips NMS8250](https://twitter.com/PanzerMSX/status/1256278384327512067) showing the same pattern as mine: 16x`00`, 16x`FF`, 16x`00`, 16x`FF`, 16x`FF`, 16x`00`, 16x`FF`, 16x`00`
* A [Sanyo Wavy 70FD](https://twitter.com/araubiMSX/status/1256302850482765824) show a `00FF00FF...` pattern ([with some suprious? bytes at the beginning](https://twitter.com/araubiMSX/status/1256299677714128896)), also not present in [its definition file](https://github.com/openMSX/openMSX/blob/master/share/machines/Sanyo_PHC-70FD.xml)
* A [Philips NMS8245](https://twitter.com/nenefranz/status/1256362447025459201) that mostly matches the pattern described in OpenMSX, except for the alternance each 256 bytes (it actually shows a `00FF00FF...` pattern).
* (...reserved for updates...)

Computers matching the OpenMSX behaviour:

* A [Mitsubishi ML-G3](https://twitter.com/PanzerMSX/status/1256287518145814529), full of `FF`s.
* A [Toshiba HX-10](https://twitter.com/PanzerMSX/status/1256290128806494209), also full of `FF`s.
* (...reserved for updates...)

Inconclusive:

* A [Philips NMS8245](https://twitter.com/PanzerMSX/status/1256285633380827136) _Note: again, I should have included a few more bytes in the tweet... this can either match the pattern described in OpenMSX, or behave as the other NMS8245 pictured above_

**Comment by theNestruo** [comment]
Hi again!

To get a wider window of memory, and a quicker overview of the differences, I've created a small ROM that basically LDIRVMs $E000-$F7FF to CHRTBL: [ram2vram.zip](https://github.com/openMSX/openMSX/files/4567401/ram2vram.zip)

This is a screenshot of OpenMSX running it in a NMS 8250:
![image](https://user-images.githubusercontent.com/12213477/80860382-99ea2300-8c67-11ea-9291-898739300d44.png)

And here is a picture of my NMS 8250 running the same ROM via MegaFlashROM cart:
![Untitled](https://user-images.githubusercontent.com/12213477/80860483-13821100-8c68-11ea-8195-a3bd0e06d1b8.jpg)

I agree this is not the best way to determine the actual _initialContents_, but I think it is a good starting point to get an overview of several machines...

**Comment by MBilderbeek** [comment]
Thanks. What I usually do is let the user use SAVEROM.COM on the RAM slot and then inspect the data.
You are right that it may depend on the used RAM chips and that they may vary from machine to machine or from batch to batch. So, it's a best effort and not 100% guaranteed correct.

I'll check out your info later.

**Comment by theNestruo** [comment]
A recap of the replies:

- Casio MX-10 ([pic](https://twitter.com/karawapo/status/1256559093927428096)): `00FF...`, alternating each 16 bytes, and skipping the alternation each 64 bytes. The pattern is not present in the XML.
- Mitsubishi ML-G3 ([pic](https://twitter.com/PanzerMSX/status/1256620432721154049)): `FF...` (correct)
- Panasonic A1GT ([pic](https://twitter.com/BeesamQ/status/1256548699955658752)): the picture shows a lot of noise (maybe the computer was not cold booted or the RAM is starting to fade out), but it seems to have a pattern of `FF00FF0000FF00FF`. Other computer (no public image available) showed a `AA` pattern with spurious bits. None of these patterns is present in the XML.
- Panasonic A1WSX ([pic](https://twitter.com/ibannieto/status/1256309325141663745) and [pic](https://twitter.com/BeesamQ/status/1256551797025538049)): one unit shows a pattern of 128x`00FF` then 128x`04FF`, while the other shows a pattern of 32x`0BFB` then 32x`0FFF`. Further investigation is required (spurious bits? faded chips?). None of these patterns is present in the XML.
- Philips VG 8235 ([pic](https://twitter.com/Skull_Erm/status/1256552864194928641) and photo in the first comment): shows a pattern of blocks of 128 bytes of `00` or `FF`. Does not match the pattern described in the XML.
- Philiips NMS 8245 ([pic](https://twitter.com/nenefranz/status/1256362447025459201) and [pic](https://twitter.com/PanzerMSX/status/1256618978618851346)): `00FF...`. Does not show the alternance each 256 bytes described in the XML.
- Philips NMS 8250 ([pic](https://twitter.com/PanzerMSX/status/1256616030895554560)): shows a pattern of blocks (each digit representing a block of 16 bytes of `00` or `FF`): `0F0F F0F0 0F0F F0F0 F0F0 0F0F F0F0 0F0F`. Seems consistent with other 8250 photos showing a larger view of the RAM (see my picture of the previous comment), but further investigation is required becuse the pattern seems to alternate in larger blocks. Does not match the pattern described in the XML.
- Sanyo Wavy 70FD ([pic](https://twitter.com/araubiMSX/status/1256302850482765824)): `00FF...`. The pattern is not present in the XML.
- Sharp HB-8000 ([pic](https://twitter.com/Fabio_acrs/status/1256656840185831427)): `FF00...00FF...`. The pattern is not present in the XML.
- Toshiba HX-10 ([pic](https://twitter.com/PanzerMSX/status/1256290128806494209)): `FF...` (correct)

Additionally: some tests on several TurboR GT and ST have shown a wide range of results: apparently random values (with predominance of `66`), mostly `AA` with spurious bits, a pattern of 256x`FF00` 256x`00FF` with spurious bits... so further investigation is required.

**Comment by MBilderbeek** [comment]
OK, for all the ones that did not have any content, I added it according to these measurements. For ones with conflicting content, I added a comment in the XML. After all, this is no exact science. The chips were not consistently used per machine type. And I don't think it's worth it to make different configs for all combinations of RAM chips used. So, let's take this as a 'best effort' :)

Thanks!
